### PR TITLE
gitserver: implement GetReposByNames batching

### PR DESCRIPTION
**Pull request by sashaostrikov**
This adds batching to already implemented [GetReposByNames function](https://github.com/sourcegraph/sourcegraph/pull/33560).

This code can probably be used as a base for a more generic batch select implementation (like [batch inserter](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/database/batch/batch.go))

## Test plan
New tests added

### CODE REVIEW

File: internal/database/gitserver_repos.go [hunk 1]
```diff
@@ -340,27 +335,70 @@ FROM gitserver_repos g
...
+
+	// iterating len(names) + 1 times because last iteration is needed for the last batch
+	for i := 0; i <= len(names); i++ {
+		if remainingNames == 0 {
+			// last batch: execute query and break out of the loop
+			if res, err := s.sendBatchQuery(ctx, maxNumPostgresParameters, nameQueries); err == nil {
+				repos = append(repos, res...)
+			} else {
+				return nil, err
+			}
```
Comment (ryanslade) [inline internal/database/gitserver_repos.go:366]: This code block is the same as the one below, I think we can always run this block and then do the checks on line 360 and 369

File: internal/database/gitserver_repos.go [hunk 2]
```diff
@@ -340,27 +335,70 @@ FROM gitserver_repos g
...
+		remainingNames--
 	}
 
-	rows, err := s.Query(ctx, sqlf.Sprintf(q, sqlf.Join(nameQueries, ",")))
+	return repos, nil
+}
+
+func (s *gitserverRepoStore) sendBatchQuery(ctx context.Context, maxNumPostgresParameters int, nameQueries []*sqlf.Query) ([]*types.GitserverRepo, error) {
+	repos := make([]*types.GitserverRepo, 0, maxNumPostgresParameters)
+	rows, err := s.Query(ctx, sqlf.Sprintf(getByNamesQueryTemplate, sqlf.Join(nameQueries, ",")))
```
Comment (ryanslade) [inline internal/database/gitserver_repos.go:389]: You should check rows.Err, using basestore.CloseRows should do what we want

File: internal/database/gitserver_repos.go [hunk 3]
```diff
@@ -340,27 +335,70 @@ FROM gitserver_repos g
...
+		batchSize++
+		remainingNames--
 	}
 
-	rows, err := s.Query(ctx, sqlf.Sprintf(q, sqlf.Join(nameQueries, ",")))
+	return repos, nil
+}
+
+func (s *gitserverRepoStore) sendBatchQuery(ctx context.Context, maxNumPostgresParameters int, nameQueries []*sqlf.Query) ([]*types.GitserverRepo, error) {
+	repos := make([]*types.GitserverRepo, 0, maxNumPostgresParameters)
```
Comment (ryanslade) [inline internal/database/gitserver_repos.go:388]: I think this is always going to create a slice with a capacity of 32767, I don't think this is what we want :)

File: internal/database/gitserver_repos.go [hunk 4]
```diff
@@ -340,27 +335,70 @@ FROM gitserver_repos g
 JOIN repo r on r.id = g.repo_id
 WHERE r.name IN (%s)
 `
-	nameQueries := []*sqlf.Query{}
-	for _, v := range names {
-		nameQueries = append(nameQueries, sqlf.Sprintf("%s", v))
+
+var MaxNumPostgresParametersMock func() int
+
+func ResetMaxNumPostgresParametersMock() {
```
Comment (ryanslade) [inline internal/database/gitserver_repos.go:341]: I know this is a pattern using in our codebase but I think we should avoid it. 

Instead we can have a private version of `GetByNames` that we use in tests and has an extra parameter for `maxNumPostgresParameters`. The public version just calls the private one with the default `batch.MaxNumPostgresParameters`

File: internal/database/gitserver_repos.go (hunk 1 above)
Comment (sashaostrikov) [reply internal/database/gitserver_repos.go:366]: the thing here is that this code block doesn't run on every loop iteration, that's why it is guarded by `if`s

For loop internals can be refactored to a such thing:
```
		if remainingNames == 0 || batchSize == maxNumPostgresParameters {
			// executing the DB query
			if res, err := s.sendBatchQuery(ctx, maxNumPostgresParameters, nameQueries); err == nil {
				repos = append(repos, res...)
			} else {
				return nil, err
			}
			
			if remainingNames == 0 {
				// last batch: break out of the loop
				break
			} else {
				// clear any variables for a new batch	
				batchSize = 0
				nameQueries = nil
			}
		}
		nameQueries = append(nameQueries, sqlf.Sprintf("%s", names[i]))
		batchSize++
		remainingNames--
```

File: internal/database/gitserver_repos.go (hunk 2 above)
Comment (sashaostrikov) [reply internal/database/gitserver_repos.go:389]: thanks, I clearly missed that one

**Comment by ryanslade** [comment]
Ping we when ready for another round of review

**Comment by sashaostrikov** [comment]
@ryanslade you were 1 minute ahead of me :)
PTAL, I've fixed all the things

**Review by asdine — APPROVED** [review]

**Review by ryanslade — APPROVED** [review]

File: internal/database/gitserver_repos.go [hunk 5]
```diff
@@ -340,17 +335,51 @@ FROM gitserver_repos g
...
+			if res, err := s.sendBatchQuery(ctx, batchSize, nameQueries); err == nil {
+				repos = append(repos, res...)
+			} else {
+				return nil, err
+			}
+
+			if remainingNames == 0 {
+				// last batch: break out of the loop
+				break
+			} else {
```
[the remark this answers is not in the document]
Comment (sashaostrikov) [reply internal/database/gitserver_repos.go:362]: fixed

**Review by ryanslade — APPROVED** [review]

**Comment by eseliger** [comment]
Out of curiosity, what is this batching for? We still collect all the results in bulk, so it can't be a memory optimization.. 🤔 Is this to keep the load on the DB more constant?

**Comment by sashaostrikov** [comment]
@eseliger well, the term "batching" is misused here, I guess. This is more of a "paging" the select query in order not to exceed pg max args count (however, this is in fact the predicate batching 😅).

But you're right, it is done so we don't run into pg limits.

**Pull request merged**